File: api-express/src/services/pedido-service.js
import * as pedidoDal from '../dal/pedido-dal.js';
import * as produtoDal from '../dal/produto-dal.js'
import AppMessage from '../utils/app-message.js'
import { mapFrete } from '../utils/tabela-frete.js';

export const create = async (payload) => {
    const skuProdutos = payload.produtos.map(p => p.skuProduto);
    payload.valorBruto = 0;
    payload.valorDesconto = 0;

    if (!skuProdutos.length) {
        throw new AppMessage('Nenhum produto informado', 400);
    }

    
    const lstProdutos = await produtoDal.getAll({
        in: skuProdutos
    }, {
        sku: true,
    })
    
    const mapProdutos = new Map();
    for (const produto of lstProdutos) {
        mapProdutos.set(produto.sku, produto);
    }
    
    if (skuProdutos.length != lstProdutos.lenght) {
        for (const sku of skuProdutos) {
            if (!mapProdutos.has(sku)) {
                throw new AppMessage(`Produto com SKU igual a "${sku}" não existe`, 400);
            }
        }
    }

    const valorFrete = mapFrete.get(Number(payload.enderecoEntrega.codigoIbgeUF));

    for (const produto of payload.produtos) {
        if (mapProdutos.get(produto.skuProduto).estoque < produto.quantidade) {
            throw new AppMessage(`Produto com SKU igual a "${produto.skuProduto}" está com estoque em falta`, 400);
        }
        
        produto.valorUnitario = mapProdutos.get(produto.skuProduto).preco;
        produto.valorBruto = produto.quantidade * produto.valorUnitario;
        produto.valorDesconto = payload.pagamento.porcentagemDesconto ? Math.round((produto.valorBruto * payload.pagamento.porcentagemDesconto) * 100) / 100 : 0;
        produto.valorLiquido = produto.valorBruto - produto.valorDesconto;
        produto.valorFrete = Math.round((valorFrete/(payload.produtos.length)) * 100) / 100;

        payload.valorBruto += produto.valorBruto;
        payload.valorDesconto += produto.valorDesconto;
    }

    // data de entrega
    let data = new Date();
    data.setDate(data.getDate() + 3);

    const dataISO = data.toISOString();
    payload.dataEntrega = dataISO;
    // fim data de entrega

    payload.valorFrete = valorFrete;
    payload.valorLiquido = payload.valorBruto - payload.valorDesconto + payload.valorFrete;

    payload.pagamento.porcentagemDesconto = undefined;
    payload.pagamento.valorTotal = payload.valorLiquido;
    payload.pagamento.valorParcela = Math.round((payload.pagamento.valorTotal/payload.pagamento.numeroParcelas) * 100) / 100;
    
    payload.enderecoEntrega.codigoIbgeUF = payload.enderecoEntrega.codigoIbgeUF.toString();
    payload.enderecoEntrega.codigoIbgeCidade = payload.enderecoEntrega.codigoIbgeCidade.toString();
    payload.status = 'P';

    const pedido = await pedidoDal.create(payload);

    for (const produto of pedido.produtos) {
        await produtoDal.updateEstoqueEVendas(produto.skuProduto, produto.quantidade * -1, produto.quantidade);
    }

    return pedido;
};

File: api-gin/src/services/pedido-service.go
package services

import (
	"math"
	"net/http"
	"tcc-api-gin/src/dal"
	"tcc-api-gin/src/models"
	"tcc-api-gin/src/utils"
	"time"
)

func CreatePedido(pedido *models.Pedido) (*models.Pedido, *utils.RestErr) {
	var skuProdutos []string
	for _, p := range pedido.Produtos {
		skuProdutos = append(skuProdutos, p.SKUProduto)
	}

	pedido.ValorBruto = 0
	pedido.ValorDesconto = 0

	if len(skuProdutos) == 0 {
		return nil, utils.NewRestErr(http.StatusBadRequest, "Nenhum produto informado", nil)
	}

	lstProdutos, restErr := dal.GetAllProdutosBySKU(skuProdutos, "*")
	if restErr != nil {
		return nil, restErr
	}

	mapProdutos := make(map[string]models.Produto)
	for _, produto := range lstProdutos {
		mapProdutos[produto.SKU] = produto
	}

	if len(skuProdutos) != len(lstProdutos) {
		for _, sku := range skuProdutos {
			if _, exists := mapProdutos[sku]; !exists {
				return nil, utils.NewRestErr(http.StatusBadRequest, "Produto com SKU '"+sku+"' não existe", nil)
			}
		}
	}

	valorFrete, exists := utils.MapFrete[pedido.Endereco.CodigoIbgeUF]
	if !exists {
		return nil, utils.NewRestErr(http.StatusBadRequest, "Código IBGE UF inválido", nil)
	}

	for i := range pedido.Produtos {
		produto := &pedido.Produtos[i]

		produtoBanco, _ := mapProdutos[produto.SKUProduto]

		// Verifica estoque
		if produtoBanco.Estoque < produto.Quantidade {
			return nil, utils.NewRestErr(http.StatusBadRequest, "Produto com SKU '"+produto.SKUProduto+"' está com estoque em falta", nil)
		}

		produto.ValorUnitario = produtoBanco.Preco
		produto.ValorBruto = produto.Quantidade * produto.ValorUnitario

		desconto := math.Round(produto.ValorBruto*(*pedido.Pagamento.PorcentagemDesconto)*100) / 100
		produto.ValorDesconto = desconto

		produto.ValorLiquido = produto.ValorBruto - produto.ValorDesconto
		produto.ValorFrete = math.Round((valorFrete/float64(len(pedido.Produtos)))*100) / 100

		pedido.ValorBruto += produto.ValorBruto
		pedido.ValorDesconto += produto.ValorDesconto
	}

	// Definir data de entrega (3 dias úteis)
	dataEntrega := time.Now().AddDate(0, 0, 3)
	pedido.DataEntrega = dataEntrega

	// Calcular valores finais do pedido
	pedido.ValorFrete = valorFrete
	pedido.ValorLiquido = pedido.ValorBruto - pedido.ValorDesconto + pedido.ValorFrete

	// Configurar pagamento
	pedido.Pagamento.PorcentagemDesconto = nil
	pedido.Pagamento.ValorTotal = pedido.ValorLiquido
	pedido.Pagamento.ValorParcela = math.Round((pedido.Pagamento.ValorTotal/float64(pedido.Pagamento.NumeroParcelas))*100) / 100

	// Definir status inicial do pedido
	pedido.Status = "P"

	// Criar pedido no banco de dados
	var pedidoResult *models.Pedido
	pedidoResult, restErr = dal.CreatePedido(pedido)
	if restErr != nil {
		return nil, restErr
	}

	// Atualizar estoque dos produtos e quantidade de vendas
	for _, produto := range pedido.Produtos {
		produtoBanco, _ := mapProdutos[produto.SKUProduto]
		produtoBanco.Estoque -= produto.Quantidade
		produtoBanco.QtdVendas += produto.Quantidade
		restErr := dal.UpdateProduto(produtoBanco)
		if restErr != nil {
			return nil, restErr
		}
	}

	return pedidoResult, nil
}
